File: frontend/types/pathWatchlist.ts
export interface PathActionItem {
  id: number;
  action_type: 'watch' | 'recheck' | 'expand' | 'alternatives' | 'archive' | 'resolve';
  metadata: Record<string, unknown> | null;
  created_at: string;
}

export interface SavedPathListItem {
  id: string;
  summary_path: string[] | null;
  path_signature: string | null;
  status: 'watching' | 'active' | 'archived' | 'resolved';
  latest_headline: string;
  recheck_count: number;
  path_length: number;
  created_at: string;
  updated_at: string;
}

export interface SavedPathDetail {
  id: string;
  path_nodes: string[];
  summary_path: string[] | null;
  path_signature: string | null;
  edge_snapshot: EdgeSnapshot[] | null;
  why_now_snapshot: WhyNowSnapshot | null;
  source_center: string | null;
  source_slot: string | null;
  status: 'watching' | 'active' | 'archived' | 'resolved';
  recheck_count: number;
  created_at: string;
  updated_at: string;
  actions: PathActionItem[];
}

export interface EdgeSnapshot {
  from: string;
  to: string;
  type: string | null;
  truth_score: number | null;
  status: string;
}

export interface WhyNowSnapshot {
  headline: string;
  signals: Array<{ type: string; count?: number; delta?: number }>;
  generated_at: string;
  strong_edges: number;
  total_edges: number;
  suggested_action?: string;
}

export interface WatchPathInput {
  path_nodes: string[];
  source_center?: string;
  source_slot?: string;
}

export interface RecheckResponse {
  headline: string;
  strengthened: EdgeDiff[];
  weakened: EdgeDiff[];
  unchanged: EdgeDiff[];
  broken_edges: EdgeDiff[];
  path_intact: boolean;
  suggested_action: string;
  suggested_reason: string;
  updated_why_now: WhyNowSnapshot;
  status: string;
  recheck_count: number;
}

export interface EdgeDiff {
  from: string;
  to: string;
  type: string | null;
  old_status: string | null;
  new_status: string | null;
  old_score: number | null;
  new_score: number | null;
}

export interface ExpandResponse {
  source_ticker: string;
  candidates: ExpandCandidate[];
  total_found: number;
}

export interface ExpandCandidate {
  ticker: string;
  name: string;
  sector: string;
  relation_type: string | null;
  truth_score: number | null;
  relation_status: string | null;
  heat_score: number | null;
  basis_summary: string;
  why_summary: string;
}

export interface AlternativesResponse {
  target_ticker: string;
  neighbor_constraints: {
    before: { ticker: string; relation_type: string } | null;
    after: { ticker: string; relation_type: string } | null;
  };
  alternatives: AlternativeCandidate[];
  total_found: number;
}

export interface AlternativeCandidate {
  ticker: string;
  name: string;
  sector: string;
  industry: string;
  overlap_count: number;
  relation_before: { rel_type: string; truth_score: number | null; status: string | null } | null;
  relation_after: { rel_type: string; truth_score: number | null; status: string | null } | null;
  why_summary: string;
}

File: apps/chain_sight/serializers/path_watchlist.py
import re

from rest_framework import serializers

from apps.chain_sight.models import PathAction, SavedPath


class PathActionSerializer(serializers.ModelSerializer):
    class Meta:
        model = PathAction
        fields = ["id", "action_type", "metadata", "created_at"]


class SavedPathListSerializer(serializers.ModelSerializer):
    latest_headline = serializers.SerializerMethodField()
    path_length = serializers.SerializerMethodField()

    class Meta:
        model = SavedPath
        fields = [
            "id",
            "summary_path",
            "path_signature",
            "status",
            "latest_headline",
            "recheck_count",
            "path_length",
            "created_at",
            "updated_at",
        ]
        read_only_fields = fields

    def get_latest_headline(self, obj) -> str:
        if obj.why_now_snapshot and isinstance(obj.why_now_snapshot, dict):
            return obj.why_now_snapshot.get("headline", "")
        return ""

    def get_path_length(self, obj) -> int:
        return len(obj.path_nodes) if obj.path_nodes else 0


class SavedPathDetailSerializer(serializers.ModelSerializer):
    actions = PathActionSerializer(many=True, read_only=True)

    class Meta:
        model = SavedPath
        fields = [
            "id",
            "path_nodes",
            "summary_path",
            "path_signature",
            "edge_snapshot",
            "why_now_snapshot",
            "source_center",
            "source_slot",
            "status",
            "recheck_count",
            "created_at",
            "updated_at",
            "actions",
        ]
        read_only_fields = [
            "id",
            "summary_path",
            "path_signature",
            "edge_snapshot",
            "why_now_snapshot",
            "status",
            "recheck_count",
            "created_at",
            "updated_at",
            "actions",
        ]


class SavedPathCreateSerializer(serializers.ModelSerializer):
    class Meta:
        model = SavedPath
        fields = ["path_nodes", "source_center", "source_slot"]

    def validate_path_nodes(self, value):
        if not isinstance(value, list):
            raise serializers.ValidationError("path_nodes는 리스트여야 합니다.")
        if len(value) < 2:
            raise serializers.ValidationError("경로는 최소 2개 노드 필요.")
        if len(value) > 10:
            raise serializers.ValidationError("경로는 최대 10개 노드까지.")
        for ticker in value:
            if not isinstance(ticker, str) or not re.match(r"^[A-Za-z]{1,5}$", ticker):
                raise serializers.ValidationError(f"유효하지 않은 ticker: {ticker}")
        if len(set(value)) != len(value):
            raise serializers.ValidationError("경로 내 중복 ticker 불가.")
        return [t.upper() for t in value]
